core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -1871,10 +1872,28 @@ && areAssignable(arg, Arrays.asList(lower, upper))) {
     switch (point.getKind()) {
     case LITERAL:
       final RexLiteral literal = (RexLiteral) point;
+      if (literal.getValue() instanceof AbstractFlatList) {
+        AbstractFlatList list = (AbstractFlatList) literal.getValue();
```

Comment:
`literal.getValueAs` will remove the cast.

Reply:
When constructing MultiSet, there is this logic that uses `FlatLists.of`, which stores `RexLiteral`. So there needs to be an additional layer of processing here, but it is not limited to `Int`.
```
      if (allowCast) {
        return makeCall(SqlStdOperatorTable.MULTISET_VALUE, operands);
      } else {
        return new RexLiteral((Comparable) FlatLists.of(operands), type,
            sqlTypeName);
      }
```

Reply:
I found using `literal.getValueAs` will return @Nullable. Additional processing may be required. So I changed it back again. Is that OK?